[REBASE & FF] ArmPkg: Add changes required for AARCH64 MSVC

VivianNK opened the pull request:
`##` Description

Add changes required for AARCH64 MSVC (Moved from old PR: https://github.com/microsoft/mu_silicon_arm_tiano/pull/355)
- Add missing functions in ArmGicV3.masm and include the .masm file in ArmGicDxe infs. 
- ArmMonitorLib.masm: change ALIGN value to fix compiler error `Compiler #2070 improper alignment value 3; expected [1,2,4,8...32768]`
- Add UINT32 ControlSystemRegValue to fix compiler type conversion warning. Previously, RegValue was being used for two values, one UINT32 and one UINT64.
- Make AllocatePageTableMemory static in CpuDxe. This fixes linker and compiler errors with ArmMmuLibPageTableAlloc. The MSVC errors were LNK2005, LNK1169, and Compiler #4744

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [x] Backport to release branch?

## How This Was Tested

Built in project consuming ArmGicV3Dxe and ArmMonitorLib

## Integration Instructions

N/A

VivianNK committed:
ArmGicV3Dxe: Add UINT32 ControlSystemRegValue
Previously, RegValue was being used for two values,
one UINT32 and one UINT64. This was causing compiler
warnings for converting between the two types.

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

os-d commented on ArmPkg/Drivers/ArmGicDxe/GicV3/ArmGicV3Dxe.c:
```diff
@@ -626,9 +627,10 @@ GicV3DxeInitialize (
+  GicV3RegValue   = ArmGicV3GetControlSystemRegisterEnable ();
+  GicV3RegValue32 = (UINT32)GicV3RegValue;
+  if ((GicV3RegValue & ICC_SRE_EL2_SRE) == 0) {
+    ArmGicV3SetControlSystemRegisterEnable (GicV3RegValue32 | ICC_SRE_EL2_SRE);
```
Why not only use `GicV3RegValue32` here? And naming it to `ControlSystemReg` or something more accurate.

```c
//UINT32
//EFIAPI
//ArmGicV3GetControlSystemRegisterEnable (
//  VOID
//  );

//VOID
//EFIAPI
//ArmGicV3SetControlSystemRegisterEnable (
//  IN UINT32         ControlSystemRegisterEnable
//  );
```

One function returns a UINT32 and the other takes a UINT32, right?

VivianNK replied:
Updated

os-d commented on ArmPkg/Drivers/ArmGicDxe/GicV3/AArch64/ArmGicV3.masm:
```diff
@@ -119,4 +121,24 @@ ArmGicV3SetBinaryPointer PROC
+//VOID
+//EFIAPI
+//ArmGicV3SetControlRegister (
+//  IN UINT32         Value
```
This is incorrect, right? It should take in a `UINTN`: https://github.com/tianocore/edk2/blob/5e5ca20bf772bc80ca71b996def6f8ebad2860bf/ArmPkg/Drivers/ArmGicDxe/GicV3/Arm/ArmGicV3.S#L90-L92

https://developer.arm.com/documentation/ddi0595/2021-06/AArch64-Registers/ICC-CTLR-EL1--Interrupt-Controller-Control-Register--EL1-

os-d commented on ArmPkg/Drivers/ArmGicDxe/GicV3/AArch64/ArmGicV3.masm:
```diff
@@ -119,4 +121,24 @@ ArmGicV3SetBinaryPointer PROC
+//EFIAPI
+//ArmGicV3GetControlRegister (
+//  VOID
+//  );
```
This is incorrect, right? Should return a UINTN? 

https://github.com/tianocore/edk2/blob/5e5ca20bf772bc80ca71b996def6f8ebad2860bf/ArmPkg/Drivers/ArmGicDxe/GicV3/Arm/ArmGicV3.S#L82-L85

https://developer.arm.com/documentation/ddi0595/2021-06/AArch64-Registers/ICC-CTLR-EL1--Interrupt-Controller-Control-Register--EL1-

os-d commented on ArmPkg/Include/Protocol/ArmPageTableMemoryAllocation.h:
```diff
@@ -34,7 +34,7 @@ VOID *
 
 typedef struct _PAGE_TABLE_MEM_ALLOC_PROTOCOL {
-  ALLOCATE_PAGE_TABLE_MEMORY    AllocatePageTableMemory;
+  ALLOCATE_PAGE_TABLE_MEMORY    AllocatePageTableMem;
```
Can you please separate the individual changes out into separate commits and REBASE & FF this? One major problem we've had with AARCH64 MSVC changes before is they are all merged together and become unwieldy and hard to deal with.

os-d replied:
Also, was renaming the field really necessary? I assume the name conflict was not for this but the function itself.

VivianNK replied:
Separated logic into 5 commits - let me know what you think

VivianNK committed:
ArmPkg: Add functions in ArmGicV3.masm and include
the .masm file in ArmGicDxe infs.

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

VivianNK committed:
ArmPkg: Change align values in .masm files to
follow standard convention.

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

VivianNK commented on ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibPageTableAlloc.c:
```diff
@@ -22,20 +22,20 @@ PAGE_TABLE_MEM_ALLOC_PROTOCOL  *mPageTableMemAllocProtocol = NULL;
 **/
 VOID *
-AllocatePageTableMemory (
+MmuAllocatePageTableMemory (
```
@os-d @kuqin12 Is this function supposed to be consumed in ArmMmuLibCore.c? Because that file has a the function declaration for `AllocatePageTableMemory` but it's not clear what its source is supposed to be. 
If this function's name is `AllocatePageTableMemory` there is a linking error: 

> error LNK2005: AllocatePageTableMemory already defined in ArmCpuDxe.lib(PageTableMemoryAllocation.obj)
INFO - ArmCpuDxe.dll.wtemp : fatal error LNK1169: one or more multiply defined symbols found

I previously made the function static but that prevents it from being consumed since it's not published as a protocol.

kuqin12 replied:
This seems like a naming conflict, did the log say which other function you are clashing with?

os-d replied:
You should be able to just make the one in CpuDxe `STATIC`, right? It looks like that is what the conflict is. The CpuDxe one should not be directly called outside of the file, only through the protocol

VivianNK replied:
Making the function static in CpuDxe worked to fix that conflict.

os-d commented on ArmPkg/Drivers/ArmGicDxe/GicV3/ArmGicV3Dxe.c:
```diff
@@ -628,7 +628,7 @@ GicV3DxeInitialize (
   RegValue = ArmGicV3GetControlSystemRegisterEnable ();
   if ((RegValue & ICC_SRE_EL2_SRE) == 0) {
-    ArmGicV3SetControlSystemRegisterEnable (RegValue | ICC_SRE_EL2_SRE);
+    ArmGicV3SetControlSystemRegisterEnable ((UINT32)RegValue | ICC_SRE_EL2_SRE);
```
This isn't needed now, right, because the next commit changes it?

VivianNK replied:
dropped

os-d commented on ArmPkg/Drivers/ArmGicDxe/GicV3/ArmGicV3Dxe.c:
```diff
@@ -688,7 +688,7 @@ GicV3DxeInitialize (
 
   // Use combined priority drop and deactivate (EOImode == 0)
-  RegValue  = ArmGicV3GetControlRegister ();
+  RegValue  = (UINT64)ArmGicV3GetControlRegister ();
```
Is this cast needed? Presumably GCC doesn't need it, if it is already checked in, and we only do MSVC for ARM64 where UINTN is UINT64, so does it actually complain without this?

VivianNK replied:
dropped

os-d commented on ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibPageTableAlloc.c:
```diff
@@ -12,7 +12,7 @@
 #include <Protocol/ArmPageTableMemoryAllocation.h>
 
-PAGE_TABLE_MEM_ALLOC_PROTOCOL  *mPageTableMemAllocProtocol = NULL;
+PAGE_TABLE_MEM_ALLOC_PROTOCOL  *mArmMmuPageTableMemAllocProtocol = NULL;
```
What does this conflict with? If you change this back, does it still give an error?

VivianNK replied:
It conflicts with the variable in 
`armpkg\drivers\cpudxe\pagetablememoryallocation.c`:
The error is that here the variable is a pointer type but in CpuDxe it's just the struct type.
`Compiler #4744 from  'mPageTableMemAllocProtocol' has different type`

os-d replied:
These can be made static right, instead of changing the name?

VivianNK replied:
That worked, yes. Updated the commit

os-d replied:
We should continue to use `STATIC`. We only used lowercase `static` for a local variable that is marked static.

VivianNK committed:
ArmPkg: Make AllocatePageTableMemory static in CpuDxe

This fixes linker and compiler errors with ArmMmuLibPageTableAlloc.
The MSVC errors were LNK2005, LNK1169, and Compiler #4744

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

VivianNK commented on ArmPkg/Drivers/CpuDxe/PageTableMemoryAllocation.c:
```diff
@@ -152,7 +153,7 @@ AllocatePageTableMemory (
 }
 
-PAGE_TABLE_MEM_ALLOC_PROTOCOL  mPageTableMemAllocProtocol = {
+static PAGE_TABLE_MEM_ALLOC_PROTOCOL  mPageTableMemAllocProtocol = {
```
@os-d We do lowercase static for variables, right?
